File: Ladder01/Python_code/permutations.py
'''Version 1: my recursion'''
class Solution:
    """
    @param nums: A list of Integers.
    @return: A list of permutations.
    """
    def permute(self, nums):
        if not nums:
            return []
        if len(nums) == 1:
            return [[nums[0]]]
        result = []
        for i in xrange(len(nums)):
            for subPermutation in self.permute(nums[:i] + nums[i + 1:]):
                result.append([nums[i]] + subPermutation)
        return result

'''Version 2: JiuZhang recursion style'''
class SolutionTwo:
    def permute(self, nums):
        if not nums:
            return []
        result = []
        self.helper(result, [], nums)
        return result

    def helper(self, result, lis, nums):
        if len(lis) == len(nums):
            result.append(lis[:])
            return
        for i in xrange(len(nums)):
            if nums[i] in lis:
                continue
            lis.append(nums[i])
            self.helper(result, lis, nums)
            lis.pop()

def main():
    test = Solution()
    nums = [3, 2, 1]
    print test.permute(nums)
if __name__ == '__main__':
    main()

File: NineChapters/01.strStr_CodingStyle/Java_Code/Permutations.java
import java.util.ArrayList;

public class Permutations {
    /**
     * @param nums: A list of integers.
     * @return: A list of permutations.
     */
    public ArrayList<ArrayList<Integer>> permute(ArrayList<Integer> nums) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        if (nums == null || nums.size() == 0) {
            return result;
        }
        ArrayList<Integer> list = new ArrayList<Integer>();
        permuteHelper(nums, result, list);
        return result;
    }

    private void permuteHelper(ArrayList<Integer> nums, ArrayList<ArrayList<Integer>> result,
        ArrayList<Integer> list) {
        if (list.size() == nums.size()) {
            result.add(new ArrayList<Integer>(list));
        }
        for (int i = 0; i < nums.size(); i++) {
            if (list.contains(nums.get(i))) {
                continue;
            }
            list.add(nums.get(i));
            permuteHelper(nums, result, list);
            list.remove(list.size() - 1);
        }
    }

    public static void main(String[] args) {
        Permutations test = new Permutations();
        ArrayList<Integer> nums = new ArrayList<Integer>();
        nums.add(1);
        nums.add(2);
        nums.add(3);
        System.out.println(test.permute(nums));
    }
}

